jerry-core/ecma/builtin-objects/ecma-builtin-regexp.c

```diff
@@ -90,36 +73,77 @@ ecma_builtin_regexp_dispatch_construct (const ecma_value_t *arguments_list_p, /*
     return regexp_value;
   }
 
-  bool is_regexp = regexp_value == ECMA_VALUE_TRUE;
+  bool pattern_is_regexp = regexp_value == ECMA_VALUE_TRUE;
+  re_compiled_code_t *bc_p = NULL;
+
+  if (new_target_p == NULL)
+  {
+    new_target_p = ecma_builtin_get (ECMA_BUILTIN_ID_REGEXP);
+
+    if (pattern_is_regexp && ecma_is_value_undefined (flags_value))
+    {
+      ecma_object_t *pattern_obj_p = ecma_get_object_from_value (pattern_value);
+
+      ecma_value_t pattern_constructor = ecma_op_object_get_by_magic_id (pattern_obj_p, LIT_MAGIC_STRING_CONSTRUCTOR);
+
+      if (ECMA_IS_VALUE_ERROR (pattern_constructor))
+      {
+        return pattern_constructor;
+      }
+
+      bool is_same = ecma_op_same_value (ecma_make_object_value (new_target_p), pattern_constructor);
+      ecma_free_value (pattern_constructor);
+
+      if (is_same)
+      {
+        return ecma_copy_value (pattern_value);
+      }
+    }
+  }
+
+  if (ecma_object_is_regexp_object (pattern_value))
+  {
+    ecma_extended_object_t *pattern_obj_p = (ecma_extended_object_t *) ecma_get_object_from_value (pattern_value);
+
+    bc_p = ECMA_GET_INTERNAL_VALUE_POINTER (re_compiled_code_t,
+                                            pattern_obj_p->u.class_prop.u.value);
+
+    if (ecma_is_value_undefined (flags_value))
+    {
+      create_regexp_from_bc = true;
+    }
+    else
+    {
+      create_regexp_from_bc = false;
+      pattern_value = bc_p->source;
+    }
+
+    pattern_is_regexp = false;
```

Comment:
Is this ok? The `pattern_is_regexp` is set to `false` if the pattern was a regexp object.

Reply:
It is set to false, because in the standard (ECMA-262 v6, 21.2.3.1) step 6 is an else if statement, so if we done the previous step, we should skip the next statement. But becasue of the backwards compatibility, i cant't use an else if statement. I couldn't come up with a better solution.

Reply:
I see, then maybe we could shuffle around things. Would this work?
```c
#if ENABLED(...) // this could be probably mered into the '#if' above
if (ecma_object_is_regexp_object (pattern_value))   // 21.2.3.1: 5. step
{
  create_regexp_from_bc = ecma_is_value_undefined (flags_value);
  if (!create_regexp_from_bc)
  { // Note that the   bc_p  assignment was moved inside this if
    ecma_extended_object_t *pattern_obj_p = (ecma_extended_object_t *) ecma_get_object_from_value (pattern_value);
    bc_p = ECMA_GET_INTERNAL_VALUE_POINTER (re_compiled_code_t,
                                          pattern_obj_p->u.class_prop.u.value);
    pattern_value = bc_p->source;
  }
}
else if (pattern_is_regexp) // 21.2.3.1: 6. step
{
  ecma_object_t *pattern_obj_p = ecma_get_object_from_value (pattern_value);
  pattern_value = ecma_op_object_get_by_magic_id (pattern_obj_p, LIT_MAGIC_STRING_SOURCE);

  if (ECMA_IS_VALUE_ERROR (pattern_value))
  {
    return pattern_value;
  }

  if (ecma_is_value_undefined (flags_value))
  {
    flags_value = ecma_op_object_get_by_magic_id (pattern_obj_p, LIT_MAGIC_STRING_FLAGS);

    if (ECMA_IS_VALUE_ERROR (flags_value))
    {
      return flags_value;
    }
  }
}
#else /* .. */
if (ecma_object_is_regexp_object (pattern_value))
{
  if (ecma_is_value_undefined (flags_value))
  {
    return ecma_copy_value (pattern_value);
  }

  return ecma_raise_type_error (ECMA_ERR_MSG ("Invalid argument of RegExp call."));
}
#endif /* .. */
```

The idea would be to clearly separate the 5.1 and 6.0 standard and it would a bit more readable. What do you thing?

Reply:
This can work, hovewer the bc_p initialization can't be moved inside that if statement, because we need it at the end of the function to create the RegExp object, if the create_regexp_from_bc is true